rust/arrow/src/datatypes.rs

```diff
@@ -151,19 +152,70 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
 make_type!(Float32Type, f32, DataType::Float32, 32, 0.0f32);
 make_type!(Float64Type, f64, DataType::Float64, 64, 0.0f64);
 
-/// A subtype of primitive type that represents numeric values.
-pub trait ArrowNumericType: ArrowPrimitiveType {}
-
-impl ArrowNumericType for Int8Type {}
-impl ArrowNumericType for Int16Type {}
-impl ArrowNumericType for Int32Type {}
-impl ArrowNumericType for Int64Type {}
-impl ArrowNumericType for UInt8Type {}
-impl ArrowNumericType for UInt16Type {}
-impl ArrowNumericType for UInt32Type {}
-impl ArrowNumericType for UInt64Type {}
-impl ArrowNumericType for Float32Type {}
-impl ArrowNumericType for Float64Type {}
+/// A subtype of primitive type that represents numeric values, if available a SIMD type and
+/// SIMD operations are defined in this trait.  SIMD is leveraged in the `compute` module if
+/// available.
+pub trait ArrowNumericType: ArrowPrimitiveType {
+    /// Defines the SIMD type that should be used for this numeric type
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    type Simd;
+
+    /// The number of SIMD lanes available
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn lanes() -> usize;
+
+    /// Loads a slice into a SIMD register
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn load(slice: &[Self::Native]) -> Self::Simd;
+
+    /// Performs a SIMD add operation
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn add(left: Self::Simd, right: Self::Simd) -> Self::Simd;
```

Comment:
Can we make this general to all math operations? e.g., +, -, *, /. Seems they are supported by the simd type.